client/db/src/utils.rs

```diff
@@ -177,41 +177,42 @@ where
 	})
 }
 
-fn backend_err(feat: &'static str) -> sp_blockchain::Error {
-	sp_blockchain::Error::Backend(feat.to_string())
-}
-
 /// Opens the configured database.
 pub fn open_database<Block: BlockT>(
-	config: &DatabaseSettings,
+	db_source: &DatabaseSource,
 	db_type: DatabaseType,
-) -> sp_blockchain::Result<Arc<dyn Database<DbHash>>> {
+	create: bool,
+) -> OpenDbResult {
 	// Maybe migrate (copy) the database to a type specific subdirectory to make it
 	// possible that light and full databases coexist
 	// NOTE: This function can be removed in a few releases
-	maybe_migrate_to_type_subdir::<Block>(&config.source, db_type).map_err(|e| {
-		sp_blockchain::Error::Backend(format!("Error in migration to role subdirectory: {}", e))
-	})?;
+	maybe_migrate_to_type_subdir::<Block>(&db_source, db_type)?;
 
-	open_database_at::<Block>(&config.source, db_type)
+	open_database_at::<Block>(&db_source, db_type, create)
 }
 
 fn open_database_at<Block: BlockT>(
 	source: &DatabaseSource,
 	db_type: DatabaseType,
-) -> sp_blockchain::Result<Arc<dyn Database<DbHash>>> {
+	create: bool,
+) -> OpenDbResult {
 	let db: Arc<dyn Database<DbHash>> = match &source {
-		DatabaseSource::ParityDb { path } => open_parity_db::<Block>(&path, db_type, true)?,
+		DatabaseSource::ParityDb { path } => open_parity_db::<Block>(&path, db_type, create)?,
 		DatabaseSource::RocksDb { path, cache_size } =>
-			open_kvdb_rocksdb::<Block>(&path, db_type, true, *cache_size)?,
-		DatabaseSource::Custom(db) => db.clone(),
+			open_kvdb_rocksdb::<Block>(&path, db_type, create, *cache_size)?,
+		DatabaseSource::Custom { db, should_init } => {
+			if *should_init && !create {
+				return Err(OpenDbError::DoesNotExist)
+			}
```

Comment:
> This error is not correct.

Please explain.

> And do we really need this check here?

That [piece](https://github.com/paritytech/substrate/blob/126dbacc2fbbe6f0ed55dd3dcb2695d7cdc71ef6/client/state-db/src/lib.rs#L563) chooses which pruning mode to use (so it will not break the "old nodes"). 

The way I implemented it relies on whether the storage had existed before it was opened.
(https://github.com/paritytech/substrate/blob/126dbacc2fbbe6f0ed55dd3dcb2695d7cdc71ef6/client/db/src/lib.rs#L1064)

That extra field in the `DatabaseSource::Custom{..}` — `should_init` helps to reproduce the same behaviour in tests:
- the newly created in-memory database does not have any meta-data entries in it — let's signal to whoever opens this `DatabaseSource` that the database is not initialised by refusing to open unless the `create` flag is set (as thought the storage had not existed before).

Reply:
> > This error is not correct.
> 
> Please explain.

If you pass in some object that implements the trait, it means the db already exists (maybe only in memory, but it exists). However, you return there that the db does not exist. Which is not true. We could add some variant "InvalidConfiguration" or whatever, but "not exist" is wrong here.

Reply:
> Maybe it would be clearer if it was named slightly differently? `require_init` or `require_create`?

`require_init` sounds better imo.

Reply:
> > > This error is not correct.
> > 
> > 
> > Please explain.
> 
> If you pass in some object that implements the trait, it means the db already exists (maybe only in memory, but it exists). However, you return there that the db does not exist. Which is not true. We could add some variant "InvalidConfiguration" or whatever, but "not exist" is wrong here.

It’s not about the Database trait, it’s about the type DatabaseSource and the “open” operation.

As I understood from the current usage the DatabaseSource::Custom variant is used solely for testing purposes.

Therefore I felt free for testing purposes introduce a piece that would help to imitate opening a non-existent database: i.e. fail with reason NotExists on attempt to open without setting the “create” flag.

Reply:
https://github.com/paritytech/substrate/blob/3dfe1b94ba84621f84dc81fa92c10c3412490248/client/db/src/lib.rs#L344-L350

I hope that name is self-explanatory enough.